boefjes/boefjes/plugins/kat_leakix/normalize.py

```diff
@@ -36,151 +43,166 @@ def run(normalizer_meta: NormalizerMeta, raw: Union[bytes, str]) -> Iterable_[OO
+        protocol=Protocol(protocol),
+        port=int(event["port"]),
+        state=PortState("open"),
+    )
+    yield ip_port_ooi
+    event_ooi = ip_port_ooi.reference
+    return event_ooi, ip_port_ooi
+
+
+def handle_hostname(event, network):
+    try:
+        ipvx = ipaddress.ip_address(event["ip"])
+        if ipvx.version == 4:
+            ip_ooi = IPAddressV4(address=event["host"], network=network)
+        else:
+            ip_ooi = IPAddressV6(address=event["host"], network=network)
+        yield ip_ooi
+        return ip_ooi.reference
+    except ValueError:
+        # Not an IPAddress, so a hostname
+        hostname_ooi = Hostname(name=event["host"], network=network)
+        yield hostname_ooi
+        return hostname_ooi.reference
+
+
+def handle_software(event, event_ooi):
+    software_args = {}
+    software_args["name"] = event.get("service", {}).get("software", {}).get("name")
+    software_args["version"] = event.get("service", {}).get("software", {}).get("version")
+    software_fingerprint = event.get("service", {}).get("software", {}).get("fingerprint")
+    if software_fingerprint:
+        software_args["name"] = software_fingerprint
+
+    if software_args["name"]:
+        software_ooi = Software(**{k: v for k, v in software_args if v})
+        yield software_ooi
+        software_instance_ooi = SoftwareInstance(ooi=event_ooi, software=software_ooi.reference)
+        yield software_instance_ooi
+        event_ooi = software_instance_ooi.reference
+        return event_ooi, software_ooi
+    return event_ooi, None
+
+
+def handle_leak(event, event_ooi, software_ooi):
+    leak_severity = event.get("leak", {}).get("severity")
+    event_source = event.get("event_source")
+    leak_stage = event.get("leak", {}).get("dataset", {}).get("stage")
+    if leak_severity or leak_stage:
+        #  Got the different severities from: https://pkg.go.dev/github.com/LeakIX/l9format#pkg-constants
+        leak_infected = event.get("leak", {}).get("dataset", {}).get("infected")
+        leak_ransomnote = event.get("leak", {}).get("dataset", {}).get("ransom_notes")
+
+        # new stage or severity, default to low
+        kat_finding = "KAT-LEAKIX-LOW"
+        if leak_severity == "critical" or leak_infected or leak_ransomnote:
+            kat_finding = "KAT-LEAKIX-CRITICAL"
+        elif leak_severity in SEVERITY_FINDING_MAPPING:
+            kat_finding = SEVERITY_FINDING_MAPPING[leak_severity]
+        elif leak_stage == "open":
+            # no severity given, default = low
+            kat_finding = "KAT-LEAKIX-LOW"
+        elif leak_stage == "explore":
+            # no severity given, default = high
+            kat_finding = "KAT-LEAKIX-HIGH"
+        elif leak_stage == "exfiltrate":
+            # no severity given, default = critical
+            kat_finding = "KAT-LEAKIX-CRITICAL"
+
+        finding_type = KATFindingType(id=kat_finding)
+        yield finding_type
+
+        kat_info = []
```

Comment:
Might as well be a concatenated string, with string starting with spaces and at the end `kat_info.lstrip()` for the end result instead of joining the list

Reply:
Joins on lists are faster, especially when the concatted string is getting longer. Ideally I'd have the list of issues as a list of issues, but wont don't support that in the Finding model just yet.

Reply:
While I understand that joins on lists might offer better performance, in this particular case, I would choose code readability over performance. The performance won't be an issue for a few literal (and formatted) strings. There's no clear winner for me in this case, so I'll leave it up to you.

Agree on extending the `Finding` model. We could do a lot more with a list of (textual) entries than just text.